### ISSUE TITLE
Incompatibility with data encrypted <=1.5.3 and later versions

### ISSUE THREAD

**Issue by Sx900** [sdrapkin/SecurityDriven.Inferno#33]
From what I can tell, if the encrypted data is larger than around 80000 bytes (maybe 81920 bytes), data encrypted with version <= 1.5.3 cannot be decrypted with versions > 1.5.3.  Similarly, data encrypted with versions > 1.5.3 cannot be decrypted by versions <= 1.5.3.

Using this code:
```c#
private byte[] DecryptTest(byte[] encryptedData, byte[] saltData, byte[] encryptionKey)
{
	ArraySegment<byte> salt = new ArraySegment<byte>(saltData);

	using (MemoryStream encryptedDataStream = new MemoryStream(encryptedData))
	{
		using (MemoryStream decryptedStream = new MemoryStream())
		{
			using (var decTransform = new EtM_DecryptTransform(key: encryptionKey, salt: salt))
			{
				using (var cryptoStream = new CryptoStream(encryptedDataStream, decTransform, CryptoStreamMode.Read))
				{
					cryptoStream.CopyTo(decryptedStream);
				}

				if (!decTransform.IsComplete) throw new Exception("Not all blocks are decrypted");
			}

			return decryptedStream.ToArray();
		}
	}
}
```

I get the following exception at the cryptoStream.CopyTo line:
```
System.Security.Cryptography.CryptographicException: Decryption failed for block 2. 
	at SecurityDriven.Inferno.EtM_DecryptTransform.TransformFinalBlock(Byte[] inputBuffer, Int32 inputOffset, Int32 inputCount) 
	at System.Security.Cryptography.CryptoStream.Read(Byte[] buffer, Int32 offset, Int32 count) 
	at System.IO.Stream.InternalCopyTo(Stream destination, Int32 bufferSize)
 ```

If the data was encrypted with version 1.5.3 and decrypted with 1.5.3, there is no error.  If the data was encrypted with version 1.6.2 and decrypted with 1.6.2, there is no error.  If the encrypted data is less than 80000 bytes, there is no error even if encrypting/decrypting across the versions.

This is using .NET 4.6.2 in case that's relevant.

I see in the commit for version 1.6.0.0

> Breaking ciphertext change:
Multi-block encryption now mixes 1st-block-salt into all subsequent blocks (thus cryptographically binding them). It prevents block #N-for-block-#N substitution from same-key multi-block encryptions where N > 1.

Is this breaking change causing my issue?  I was not using the chunkNumber parameter for EtM_EncryptTransform / EtM_DecryptTransform constructors.   Do I need to decrypt all data using version 1.5.3 and then re-encrypt it using version 1.6.2 for it to be decryptable using version 1.6.2?  Or just stick with version 1.5.3 indefinitely?

**Comment by sdrapkin** [comment]
@Sx900 

> Is this breaking change causing my issue?

Correct. There was a insecurity in how multi-chunk encryption was implemented under same-key scenario (chunk-N from file-1 could replace chunk-N from file-2, yet still validate). This has been fixed in v1.6.0.0, but is a breaking change in the multi-chunk usecase. Very low risk of attack-vector, unless it's a special case.

>  Do I need to decrypt all data using version 1.5.3 and then re-encrypt it using version 1.6.2 for it to be decryptable using version 1.6.2? Or just stick with version 1.5.3 indefinitely?

You could keep using v1.5.3 indefinitely. However - if you're able to - a better solution is to do what you said: decrypt with v1.5.3 and then re-encrypt with vLatest.

I apologize for the breaking change - had no choice in order to fix.

**Comment by Sx900** [comment]
Okay - thanks for the quick response.  I'll look into doing a decrypt/re-encrypt conversion.

Any chance the "Releases" section (https://github.com/sdrapkin/SecurityDriven.Inferno/releases) could be updated to include each Nuget release (the last one listed is 1.5.3)?  The 1.6.0 release notes could highlight the breaking change.
